controllers/src/main/java/com/epam/esm/controllers/GiftCertificateController.java

```diff
@@ -0,0 +1,132 @@
+            throw new ResourceNotFoundException(resource);
+        }
+    }
+
+    /**
+     * Finds giftCertificate by id and creates a corresponding dto object;
+     *
+     * @param id is a unique field of giftCertificate;
+     * @return GiftCertificateDto is the generated corresponding dto object based on the found giftCertificate
+     * @throws ResourceNotFoundException if the requested giftCertificate is not found;
+     */
+    @GetMapping("{id}")
+    public GiftCertificateDto getGiftCertificate(@PathVariable Long id) throws ResourceNotFoundException {
+        Optional<GiftCertificateDto> giftCertificateDto = giftCertificateService.findById(id);
+        if (giftCertificateDto.isPresent()) {
+            return giftCertificateDto.get();
+        } else {
+            String resource = " (Gift certificate id = " + id + ")";
+            throw new ResourceNotFoundException(resource);
+        }
+    }
+
+    /**
+     * Deletes giftCertificate by its id;
+     *
+     * @param id is a unique field of giftCertificate;
+     * @throws ResourceNotFoundException if the resource being deleted does not found;
+     */
+    @DeleteMapping("{id}")
+    public void deleteGiftCertificate(@PathVariable Long id) throws ResourceNotFoundException {
+        Optional<GiftCertificateDto> giftCertificate = giftCertificateService.findById(id);
+        if (giftCertificate.isPresent()) {
+            giftCertificateService.deleteGiftCertificate(id);
+        } else {
+            String resource = " (Gift certificate id " + id + ")";
+            throw new ResourceNotFoundException(resource);
+        }
+    }
+
+    /**
+     * Finds giftCertificates by name of the associated tag and creates a corresponding dto objects;
+     *
+     * @param tagName is a unique name of tag;
+     * @return a collection <code>List</code> contains the GiftCertificateDto objects
+     * or empty collection <code>List</code>;
+     */
+    @GetMapping("tag/{name}")
+    public List<GiftCertificateDto> findCertificatesByTag(@PathVariable("name") String tagName) {
+        return giftCertificateService.findGiftCertificatesByTag(tagName);
+    }
+
+    /**
+     * Finds giftCertificates by match in name or description and creates a corresponding dto objects;
+     *
+     * @param searchCondition is a part and whole word that may appear in the name or description
+     *                        of giftCertificate;
+     * @return a collection <code>List</code> contains the GiftCertificateDto objects
+     * or empty collection <code>List</code>;
+     */
+    @GetMapping("search/{condition}")
+    public List<GiftCertificateDto> findCertificatesByNameOrDescription(@PathVariable("condition") String searchCondition) {
+        return giftCertificateService.findGiftCertificatesByNameOrDescription(searchCondition);
+    }
+
+    /**
+     * Finds giftCertificates by concrete sort condition and creates a corresponding dto objects;
```

Comment:
for me is not clear what string should I pass in case I want to sort certificates by date descending, so the endpoint is not documented completely. 

[optionally] Also it would be great if I have an option to search most recent tags by part of the tag name